services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py

```diff
@@ -0,0 +1,104 @@
+import logging
+import time
+from collections.abc import Awaitable, Callable
+from datetime import timedelta
+from typing import Final
+
+from pydantic.errors import PydanticErrorMixin
+from tenacity import AsyncRetrying, RetryCallState, TryAgain
+from tenacity.stop import stop_after_delay
+from tenacity.wait import wait_fixed
+
+_logger = logging.getLogger(__name__)
+
+
+_DEFAULT_CHECK_INTERVAL: Final[timedelta] = timedelta(seconds=1)
+_DEFAULT_TIMEOUT_INTERVAL: Final[timedelta] = timedelta(seconds=30)
+
+
+class CouldNotReachServiceError(PydanticErrorMixin, Exception):
+    msg_template: str = "Could not contact service '{service_name}' at '{endpoint}'. Look above for details."
+
+
+def _before_sleep_log(
+    logger: logging.Logger, service_name: str, endpoint: str
+) -> Callable[[RetryCallState], None]:
+    def log_it(retry_state: RetryCallState) -> None:
+        assert retry_state  # nosec
+        assert retry_state.next_action  # nosec
+        logger.warning(
+            "Retrying to contact '%s' at '%s' in %s seconds.",
+            service_name,
+            endpoint,
+            retry_state.next_action.sleep,
+        )
+
+    return log_it
+
+
+async def _attempt_to_wait_for_handler(
+    async_handler: Callable[..., Awaitable],
+    *args,
+    service_name: str,
+    endpoint: str,
+    check_interval: timedelta,
+    timeout: timedelta,
+    **kwargs,
+) -> None:
+    async for attempt in AsyncRetrying(
+        wait=wait_fixed(check_interval),
+        stop=stop_after_delay(timeout.total_seconds()),
+        before_sleep=_before_sleep_log(_logger, service_name, endpoint),
+        reraise=True,
+    ):
+        with attempt:
```

Comment:
minor: this form is bad as it does not show in the log which function is retried. better to use the decorator.